File: Library/PaintedSkin/src/main/java/org/alee/component/skin/parser/ThemeSkinExecutorBuilderManager.java
package org.alee.component.skin.parser;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

import org.alee.component.skin.executor.ISkinExecutor;
import org.alee.component.skin.executor.SkinElement;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**********************************************************
 *
 * @author: MY.Liu
 * @date: 2021/2/14
 * @description: xxxx
 *
 *********************************************************/
@RestrictTo(RestrictTo.Scope.LIBRARY)
public final class ThemeSkinExecutorBuilderManager {
    /**
     * {@link IThemeSkinExecutorBuilder}
     */
    private final List<IThemeSkinExecutorBuilder> mThemeSkinExecutorBuilders;
    /**
     * {@link ThreadLocal}
     */
    private final ThreadLocal<List<IThemeSkinExecutorBuilder>> mThreadLocal;

    private ThemeSkinExecutorBuilderManager() {
        mThemeSkinExecutorBuilders = new ArrayList<>();
        mThreadLocal = new ThreadLocal<List<IThemeSkinExecutorBuilder>>() {
            @Nullable
            @Override
            protected List<IThemeSkinExecutorBuilder> initialValue() {
                return new ArrayList<>();
            }
        };
    }

    /**
     * 获取单例对象
     *
     * @return {@link ThemeSkinExecutorBuilderManager}
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public static ThemeSkinExecutorBuilderManager getInstance() {
        return ThemeSkinExecutorBuilderManagerHolder.INSTANCE;
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public void addThemeSkinExecutorBuilder(@NonNull IThemeSkinExecutorBuilder builder) {
        synchronized (mThemeSkinExecutorBuilders) {
            if (mThemeSkinExecutorBuilders.contains(builder)) {
                return;
            }
            mThemeSkinExecutorBuilders.add(builder);
        }
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public void removeThemeSkinExecutorBuilder(IThemeSkinExecutorBuilder builder) {
        synchronized (mThemeSkinExecutorBuilders) {
            mThemeSkinExecutorBuilders.remove(builder);
        }
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public SkinElement[] parseElement(Context context, AttributeSet attributeSet) {
        if (null == context || null == attributeSet || 0 >= attributeSet.getAttributeCount()) {
            return null;
        }
        List<IThemeSkinExecutorBuilder> tempList = mThreadLocal.get();
        synchronized (mThemeSkinExecutorBuilders) {
            tempList.clear();
            tempList.addAll(mThemeSkinExecutorBuilders);
        }
        Set<SkinElement> tempSet = new HashSet<>();
        for (int i = tempList.size() - 1; i >= 0; i--) {
            IThemeSkinExecutorBuilder temp = tempList.get(i);
            if (null == temp) {
                continue;
            }
            Set<SkinElement> supportElement = temp.parse(context, attributeSet);
            if (null == supportElement || supportElement.isEmpty()) {
                continue;
            }
            tempSet.addAll(supportElement);
        }
        return tempSet.toArray(new SkinElement[tempSet.size()]);
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public void appendAttr(Context context, SkinElement element) {
        element.setResourcesName(context.getResources().getResourceEntryName(element.getResourcesId()));
        element.setResourcesType(context.getResources().getResourceTypeName(element.getResourcesId()));
    }

    public boolean isValidResources(int resourcesId) {
        if (0 >= resourcesId) {
            return false;
        }
        return !Integer.toHexString(resourcesId).startsWith("1");
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public ISkinExecutor obtainSkinExecutor(View view, SkinElement element) {
        if (null == view || null == element) {
            return null;
        }
        List<IThemeSkinExecutorBuilder> tempList = mThreadLocal.get();
        synchronized (mThemeSkinExecutorBuilders) {
            tempList.clear();
            tempList.addAll(mThemeSkinExecutorBuilders);
        }
        for (int i = tempList.size() - 1; i >= 0; i--) {
            IThemeSkinExecutorBuilder temp = tempList.get(i);
            if (null == temp) {
                continue;
            }
            if (!temp.isSupportAttr(view, element.getAttrName())) {
                continue;
            }
            ISkinExecutor executor = temp.requireSkinExecutor(view, element);
            if (null == executor) {
                continue;
            }
            return executor;
        }
        return null;
    }

    /**
     * 静态内部类持有外部对象实现单利方式
     */
    private static class ThemeSkinExecutorBuilderManagerHolder {
        /**
         * {@link ThemeSkinExecutorBuilderManager}
         */
        private static final ThemeSkinExecutorBuilderManager INSTANCE = new ThemeSkinExecutorBuilderManager();
    }
}

File: Library/PaintedSkin-Kotlin/src/main/java/org/alee/component/skin/core/parser/ThemeSkinExecutorBuilderManager.kt
package org.alee.component.skin.core.parser

import android.content.Context
import android.util.AttributeSet
import android.view.View
import org.alee.component.skin.model.SkinAttribute
import org.alee.component.skin.template.ISkinExecutor
import org.alee.component.skin.template.ISkinExecutorBuilder
import org.alee.component.skin.util.SafeArray

/**
 * 换肤执行器的构造器管理类
 *
 * <p> 详细描述
 *
 * @author MingYu.Liu
 * created in 2023/8/31
 *
 */
internal object ThemeSkinExecutorBuilderManager {
    /**
     * [ISkinExecutorBuilder] 持有者
     */
    private val mSkinExecutorBuilders by lazy { SafeArray<ISkinExecutorBuilder>() }

    /**
     * 添加一个[ISkinExecutorBuilder]
     * @param builder ISkinExecutorBuilder
     */
    fun addSkinExecutorBuilder(builder: ISkinExecutorBuilder) {
        mSkinExecutorBuilders.add(builder)
    }

    /**
     * 移除一个[ISkinExecutorBuilder]
     * @param builder ISkinExecutorBuilder
     */
    fun removeSkinExecutorBuilder(builder: ISkinExecutorBuilder) {
        mSkinExecutorBuilders.remove(builder)
    }

    /**
     * 从View的属性集中解析出支持换肤的属性
     * @param context Context [Context]
     * @param attributeSet AttributeSet [AttributeSet]
     * @return Set<SkinAttribute> [SkinAttribute]
     */
    fun fetchSkinAttribute(context: Context, attributeSet: AttributeSet): Set<SkinAttribute> {
        if (attributeSet.isEmpty()) {
            return emptySet()
        }
        mSkinExecutorBuilders.use().use {
            return it.mapNotNull { builder -> builder.parse(context, attributeSet) }.flatten().toSet()
        }
    }

    /**
     * 根据View与属性构建换肤执行器
     * @param view View 需要换肤的View
     * @param skinAttribute SkinAttribute 需要换肤的属性
     * @return ISkinExecutor? [ISkinExecutor]
     */
    fun obtainSkinExecutor(view: View, skinAttribute: SkinAttribute): ISkinExecutor? {
        mSkinExecutorBuilders.use().use {
            return it.firstOrNull { builder -> builder.canBuildExecute(view, skinAttribute.attrName) }
                ?.buildSkinExecutor(view, skinAttribute)
        }
        return null
    }

    private fun AttributeSet.isEmpty(): Boolean {
        return 0 >= attributeCount
    }
}
